### PR TITLE
FIX: allow start-stop subplot

### LINKED ISSUE #17343: Regression in add_subplot..
https://github.com/matplotlib/matplotlib/pull/16527 419fafeac34c added checks for `add_subplot` that break:

```python
import matplotlib.pyplot as plt

fig = plt.figure()

ax = fig.add_subplot(3, 4, 1)
ax2 = fig.add_subplot(3, 2, (3, 5))
plt.show()
```

### Old behaviour

Figure_1

### New behaviour:

```
  ax2 = fig.add_subplot(3, 2, (3, 5))
Traceback (most recent call last):
  File "testPlots.py", line 6, in <module>
    ax2 = fig.add_subplot(3, 2, (3, 5))
  File "/Users/jklymak/matplotlib/lib/matplotlib/figure.py", line 1376, in add_subplot
    args = tuple(map(int, args))
TypeError: int() argument must be a string, a bytes-like object or a number, not 'tuple'
```

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #17343

`add_subplot` is meant to accept spans for the third parameter, but this was broken by #16527 which was doing more careful argument checking.   Happy to take criticism on how I did this, but I think it works...

```python 
import matplotlib.pyplot as plt

fig = plt.figure()

ax = fig.add_subplot(3, 4, 1)
ax2 = fig.add_subplot(3, 2, (3, 5))
plt.show()
```

Figure_1

TODO:

- [x] Add tests so this doesn't break again
- [ ] Add examples?
- [x] Fix the docstring to make this clear

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by tacaswell** [comment]
I'm with @story645 's comment in the original issue "we support that ?!?".  Agree that we should not break it.

File: lib/matplotlib/figure.py [hunk 1]
```diff
@@ -1264,13 +1264,16 @@ def add_subplot(self, *args, **kwargs):
...
-        *args, int or (int, int, int) or `SubplotSpec`, default: (1, 1, 1)
+        *args, int, (int, int, *index*), or `SubplotSpec`, default: (1, 1, 1)
             The position of the subplot described by one of
 
             - Three integers (*nrows*, *ncols*, *index*). The subplot will
               take the *index* position on a grid with *nrows* rows and
               *ncols* columns. *index* starts at 1 in the upper left corner
-              and increases to the right.
+              and increases to the right. *index* can also be  a two-tuple
+              sepcifying the (*start*, *stop*) indices of the subplot.  i.e.
```
Comment (QuLogic) [inline lib/matplotlib/figure.py:1274]:
```suggestion
              specifying the (*start*, *stop*) indices of the subplot, i.e.,
```

**Comment by jklymak** [comment]
matlab documentation: https://www.mathworks.com/help/matlab/ref/subplot.html  See the subhead under "p" 

> Grid position for the new axes, specified as a scalar or vector of positive integers.
>
> If p is a scalar positive integer, then subplot creates a subplot in grid position p.
> If p is a vector of positive integers, then subplot creates a subplot that spans the grid positions listed in p.
> Example: subplot(2,3,1) creates a subplot in position 1.
> 
> Example: subplot(2,3,[2,5]) creates a subplot spanning positions 2 and 5.
>
> Example: subplot(2,3,[2,6]) creates a subplot spanning positions 2, 3, 5, and 6.

File: lib/matplotlib/figure.py [hunk 2]
```diff
@@ -1377,14 +1380,24 @@ def add_subplot(self, *args, **kwargs):
...
+                    cbook.warn_deprecated("3.3", message=message)
+                newarg[nn] = int(arg)
+            if isinstance(args[2], tuple):
+                # start/stop two-tuple is allowed...
+                for arg in args[2]:
+                    if not isinstance(arg, Integral):
+                        cbook.warn_deprecated("3.3", message=message)
+                newarg[nn] = (int(args[0]), int(args[1]))
+            else:
+                newarg[nn] = int(arg)
```
Comment (QuLogic) [inline lib/matplotlib/figure.py:1398]: Missing a warning if this is not an int.

File: lib/matplotlib/figure.py [hunk 3]
```diff
@@ -1377,14 +1380,24 @@ def add_subplot(self, *args, **kwargs):
...
-                                "%(since)s and will be removed %(removal)s.")
-            args = tuple(map(int, args))
+                    cbook.warn_deprecated("3.3", message=message)
+                newarg[nn] = int(arg)
+            if isinstance(args[2], tuple):
+                # start/stop two-tuple is allowed...
+                for arg in args[2]:
+                    if not isinstance(arg, Integral):
+                        cbook.warn_deprecated("3.3", message=message)
+                newarg[nn] = (int(args[0]), int(args[1]))
```
Comment (QuLogic) [inline lib/matplotlib/figure.py:1396]: Needs a length check then, as it accepts `(1, 2, 3, 4)` and drops the last two.

Comment (jklymak) [reply lib/matplotlib/figure.py]: oops, yeah, and it was wrong too..

Comment (jklymak) [reply lib/matplotlib/figure.py]: eww, very wrong. Sorry about that.  I think its fixed now.  But checking this is a really ugly mess...  i.e. I'm not convinced the cure isn't worse than the disease.

File: lib/matplotlib/figure.py [hunk 4]
```diff
@@ -1377,14 +1380,26 @@ def add_subplot(self, *args, **kwargs):
...
                 if not isinstance(arg, Integral):
-                    cbook.warn_deprecated(
-                        "3.3",
-                        message="Passing non-integers as three-element "
-                                "position specification is deprecated since "
-                                "%(since)s and will be removed %(removal)s.")
-            args = tuple(map(int, args))
+                    cbook.warn_deprecated("3.3", message=message)
+                newarg[nn] = int(arg)
+            args2 = args[2]
```
Comment (QuLogic) [inline lib/matplotlib/figure.py:1391]: Unused... I assume this should replace all `args[2]` below, though.

Comment (jklymak) [reply lib/matplotlib/figure.py:1391]: I really shouldn't do this stuff while also watching the kids...

File: lib/matplotlib/figure.py [hunk 5]
```diff
@@ -1264,13 +1264,16 @@ def add_subplot(self, *args, **kwargs):
...
-        *args, int or (int, int, int) or `SubplotSpec`, default: (1, 1, 1)
+        *args, int, (int, int, *index*), or `SubplotSpec`, default: (1, 1, 1)
             The position of the subplot described by one of
 
             - Three integers (*nrows*, *ncols*, *index*). The subplot will
               take the *index* position on a grid with *nrows* rows and
               *ncols* columns. *index* starts at 1 in the upper left corner
-              and increases to the right.
+              and increases to the right. *index* can also be  a two-tuple
+              specifying the (*start*, *stop*) indices of the subplot, i.e.,
```
Comment (QuLogic) [inline lib/matplotlib/figure.py:1274]: Oops, just realized this is an e.g., not i.e..
```suggestion
              specifying the (*start*, *stop*) indices of the subplot, e.g.,
```

**Comment by anntzer** [comment]
See https://github.com/matplotlib/matplotlib/pull/17350 for an alternative approach.

**Comment by jklymak** [comment]
Closing in lieu of #17350 (though I would somewhat prefer to get a request to revise than a completely different PR that does the same thing).

**Comment by anntzer** [comment]
I generally try to do that, but here the whole approach was so different that I though it would be faster to just write it rather than to describe what I'd prefer.  Sorry about that, wasn't trying to PR-snipe you :-)

**Pull request closed without merging**